src/Cardano/Api/Types.hs

```diff
@@ -0,0 +1,283 @@
+
+newtype Amount = Amount Natural
+    deriving stock (Generic, Show, Ord, Eq)
+
+instance FromJSON Amount where
+    parseJSON x = Amount . removeUnit @"lovelace" <$> parseJSON x
+
+instance ToJSON Amount where
+    toJSON (Amount x) = toJSON $ addUnit @"lovelace" x
+
+newtype Percentage = Percentage Int
+    deriving stock (Eq, Generic, Ord, Show)
+
+data PercentageError
+    = PercentageOutOfBoundsError
+    deriving (Show)
+
+instance Bounded Percentage where
+    minBound = Percentage 0
+    maxBound = Percentage 100
+
+instance FromJSON Percentage where
+    parseJSON x = eitherToParser
+        . mkPercentage @Int . removeUnit @"percent" =<< parseJSON x
+
+instance ToJSON Percentage where
+    toJSON (Percentage x) = toJSON $ addUnit @"percent" x
+
+mkPercentage :: Integral i => i -> Either PercentageError Percentage
+mkPercentage i
+    | j < minBound = Left PercentageOutOfBoundsError
+    | j > maxBound = Left PercentageOutOfBoundsError
+    | otherwise = pure j
+  where
+    j = Percentage (fromIntegral i)
+
+data Wallet = Wallet
+    { _id :: !WalletId
+    , _addressPoolGap :: !(ApiT AddressPoolGap)
+    , _balance :: !WalletBalance
+    , _delegation :: !WalletDelegation
+    , _name :: !(ApiT WalletName)
+    , _passphrase :: !WalletPassphraseInfo
+    , _state :: !WalletState
+    } deriving (Eq, Generic, Show)
+
+instance FromJSON Wallet where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance ToJSON Wallet where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+instance ToJSON (ApiT WalletName) where
+    toJSON = toJSON . getWalletName . getApiT
+instance FromJSON (ApiT WalletName) where
+    parseJSON x = fmap ApiT . eitherToParser . mkWalletName =<< parseJSON x
+
+deriving newtype instance Bounded (ApiT AddressPoolGap)
+deriving newtype instance Enum (ApiT AddressPoolGap)
+instance FromJSON (ApiT AddressPoolGap) where
+    parseJSON x = do
+        gap <- parseJSON x
+        ApiT <$> eitherToParser (mkAddressPoolGap gap)
+instance ToJSON (ApiT AddressPoolGap) where
+    toJSON = toJSON . getAddressPoolGap . getApiT
+
+data WalletBalance = WalletBalance
+    { _available :: !Amount
+    , _total :: !Amount
+    } deriving (Eq, Generic, Show)
+
+instance FromJSON WalletBalance where
+    parseJSON = genericParseJSON defaultRecordTypeOptions
+instance ToJSON WalletBalance where
+    toJSON = genericToJSON defaultRecordTypeOptions
+
+-- | Wallet Delegation representation, can be serialized to and from JSON as
+-- follows:
+--
+-- >>> Aeson.encode NotDelegating
+-- {"status":"not_delegating"}
+--
+-- >>> Aeson.encode $ Delegating poolId
+-- {"status":"delegating","target": "27522fe5-262e-42a5-8ccb-cef884ea2ba0"}
+data WalletDelegation
+    = NotDelegating
+    | Delegating !UUID
+    deriving (Eq, Generic, Show)
+
+instance FromJSON WalletDelegation where
+    parseJSON = genericParseJSON walletDelegationOptions
+instance ToJSON WalletDelegation where
+    toJSON = genericToJSON walletDelegationOptions
```

Comment:
This is just my opinion, but for small types like this I think it's more straightforward to write instances manually rather than set up the options for `genericToJSON`.

Reply:
I agree, especially in the case where the particular set of options is used just once. Perhaps we can tidy this up a bit in a future PR.

We're going to be adding quite a lot more to the API modules, so there will be lots of opportunities to do some tidying on the way.